### PR TITLE
Add new rule `no-shadow-route-definition`

### PR DESCRIPTION
**Pull request by raido**
This PR adds new rule to avoid route path definition collisions for Router.

Example message: `Route "blog" (/blog, 3L:8C) is shadowing route "blog" (/blog, 2L:1C)`

Closes: #951

### CODE REVIEW

File: tests/lib/rules/no-shadow-route-definition.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      this.route("post", { path: "/" });
+    });
+    this.route("post")`,
+    `this.route("blog", function() {
+      this.route("post", { path: "/" });
+    });
+    this.route("post")`,
+
+    // With dynamic segment:
+    'this.route("blog", { path: ":blog" });',
```
Comment (raido) [inline tests/lib/rules/no-shadow-route-definition.js:59]: Actually this is incorrect for this rule, as both path definitions would equal `blog/<value>` 🤔 . I think I need to work a bit more on this.

**Comment by raido** [comment]
This is now ready for feedback :)

File: tests/lib/rules/no-shadow-route-definition.js [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+  invalid: [
+    {
+      code: `
+        this.route("blog");
+        this.route("blog");
+      `,
+      output: null,
+      errors: [
+        {
+          message: buildErrorMessage({
```
Comment (bmish) [inline tests/lib/rules/no-shadow-route-definition.js:119]: Could we check the string error message in at least one test? This `buildErrorMessage` is pretty complicated and we actually need to test that that generates the right string output too.

Comment (raido) [reply tests/lib/rules/no-shadow-route-definition.js:126]: Test added.

File: tests/lib/rules/no-shadow-route-definition.js [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    "test('blog', { path: '/blog' });",
+    "this.test('blog');",
+    "this.test('blog', { path: 'blog' });",
+    "this.test('blog', { path: '/blog' });",
+    "MyClass.route('blog');",
+    "MyClass.route('blog', { path: 'blog' });",
+    "MyClass.route('blog', { path: '/blog' });",
+    "route.unrelatedFunction('blog', { path: 'blog' });",
+    "this.route.unrelatedFunction('blog', { path: 'blog' });",
+  ],
```
[the remark this answers is not in the document]
Comment (raido) [reply tests/lib/rules/no-shadow-route-definition.js:116]: Test added.

File: lib/rules/no-shadow-route-definition.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    const pathOptionNode = getPropertyByKeyName(node.arguments[1], 'path');
+    const pathValue = pathOptionNode.value.value;
+    if (pathValue) {
+      routePath = pathValue;
+    }
+  }
+  return routePath;
+}
+
+function getParentRoutesPaths(node, stack = []) {
```
[the remark this answers is not in the document]
Comment (raido) [reply lib/rules/no-shadow-route-definition.js]: Refactored.

**Comment by raido** [comment]
I addressed feedback but this PR needs more work. Wildcard and params route detection logic is flawed.

Example:

```
this.route("first", { path: "/:first" });
this.route("second", { path: "/:second" });
```

Would not be detected as conflict but from router perspective it is. Just parameter name is different but URLs would be the same. Until I have fixed this, this cannot be merged.

**Comment by raido** [comment]
I've implemented a solution for matching params and wildcards.

**Review by bmish — APPROVED** [review]

**Pull request merged**